docs/devel/controllers.md

```diff
@@ -0,0 +1,101 @@
+# Writing Controllers
+
+A Kubernetes controller is simply an optimization on bash script that does repeated lists against the API server.
+
+## Guidelines
+
+When you’re writing controllers, there are few guidelines that will help make sure you get the results and performance you’re looking expecting.
+
+1. Operate on one item at a time.  If you use a `workqueue.Interface`, you’ll be queue changes for a particular resource and later pop them in multiple “worker” gofuncs with a guarantee that no two gofuncs will work on the same item at the same time.
+
+ Many controllers must trigger off multiple resources (I need to check X if Y changes), but nearly all controllers can collapse those into a queue of “check this X” based on relationships.  For instance, a ReplicaSetController needs to react to a pod being deleted, but it does that by finding the related ReplicaSets and queuing those. 
+
+
+1. Random ordering between resources. When controllers queue off multiple types of resources, there is no guarantee of ordering amongst those resources.
+
+ Distinct watches are updated independently.  Even with an objective ordering of “created resourceA/X” and “created resourceB/Y”, your controller could observe “created resourceB/Y” and “created resourceA/X”.
+
+
+1. Level driven, not edge driven.  Just like having a bash script that isn’t running all the time, your controller may off for an indeterminate amount of time before running again.
+
+ If an API object appears with a marker value of `true`, you can’t count on having seen it turn from `false` to `true`, only that you now observe it being `true`.  Even an API watch suffers from this problem, so be sure that you’re not counting on seeing a change unless your controller is also marking the information it last made the decision on in the object’s status.
```

Comment:
"is also marking the information it last made the decision on in the object’s status." - I wished the English language had some more commas such that you don't need a non-regular grammar to parse it.